contracts/ColonyNetworkAuction.sol

```diff
@@ -0,0 +1,216 @@
+  (at your option) any later version.
+
+  The Colony Network is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with The Colony Network. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+pragma solidity ^0.4.17;
+pragma experimental "v0.5.0";
+pragma experimental "ABIEncoderV2";
+
+import "../lib/dappsys/math.sol";
+import "./ColonyNetworkStorage.sol";
+import "./ERC20Extended.sol";
+import "./IColony.sol";
+
+
+contract ColonyNetworkAuction is ColonyNetworkStorage {
+  event AuctionCreated(address auction, address token, uint256 quantity);
+
+  function startTokenAuction(address _token) public {
+    address commonColony = _colonies["Common Colony"];
+    address clny = IColony(commonColony).getToken();
+    DutchAuction auction = new DutchAuction(clny, _token);
+    uint availableTokens = ERC20Extended(_token).balanceOf(this);
+    ERC20Extended(_token).transfer(auction, availableTokens);
+    AuctionCreated(address(auction), _token, availableTokens);
+  }
+}
+
+
+contract DutchAuction is DSMath {
+  address public colonyNetwork;
+  ERC20Extended public clnyToken;
+  ERC20Extended public token;
+
+  // Total number of auctioned tokens
+  uint public quantity;
+  bool public started;
+  uint public startTime;
+  uint public endTime;
+  uint public constant TOKEN_DECIMALS = 10 ** 18;
+  
+  // Keep track of all CLNY wei received
+  uint public receivedTotal;
+  uint public bidCount;
+  uint public claimCount;
+
+  // Final price in CLNY per 10**18 Tokens (min 1, max 1e18)
+  uint public finalPrice;
+  bool public finalized;
+  
+  mapping (address => uint256) public bids;
+
+  modifier auctionNotStarted {
+    require(startTime == 0);
+    require(!started);
+    _;
+  }
+
+  modifier auctionStartedAndOpen {
+    require(started);
+    require(startTime > 0);
+    require(endTime == 0);
+    _;
+  }
+
+  modifier auctionClosed {
+    require(endTime > 0);
+    _;
+  }
+
+  modifier auctionNotFinalized() {
+    require(!finalized);
+    _;
+  }
+
+  modifier auctionFinalized {
+    require(finalized);
+    _;
+  } 
+
+  modifier allBidsClaimed  {
+    require(claimCount == bidCount);
+    _;
+  }
+
+  event AuctionBid(address indexed _sender, uint _amount, uint _missingFunds);
+  event AuctionClaim(address indexed _recipient, uint _sentAmount);
+  event AuctionFinalized(uint _finalPrice);
+
+  function DutchAuction(address _clnyToken, address _token) public {
+    colonyNetwork = msg.sender;
+    require(_clnyToken != 0x0 && _token != 0x0);
+    clnyToken = ERC20Extended(_clnyToken);
+    token = ERC20Extended(_token);
+  }
+
+  function start() public
+  auctionNotStarted
+  {
+    quantity = token.balanceOf(this);
+    // Quantity has to be a minimum of 1e18 tokens
```

Comment:
On 1) it's because we open `totalToEndAuction` to be 0 in cases where `quantity(of tokens sold) * price(clny per 1e18 tokens) < 1e18`, e.g. if `price = 1` and `quantity < 1e18`

Reply:
ok so what I'll do instead is maintain a variable (per auction) minimum price = `1e18/quantity`